gossip/src/duplicate_shred.rs

```diff
@@ -84,35 +85,71 @@ pub enum Error {
     TryFromIntError(#[from] TryFromIntError),
     #[error("unknown slot leader: {0}")]
     UnknownSlotLeader(Slot),
+    #[error("invalid last index conflict")]
+    InvalidLastIndexConflict,
+    #[error("invalid erasure meta conflict")]
+    InvalidErasureMetaConflict,
 }
 
-// Asserts that the two shreds can indicate duplicate proof for
-// the same triplet of (slot, shred-index, and shred-type_), and
-// that they have valid signatures from the slot leader.
+/// Check that `shred1` and `shred2` indicate a valid duplicate proof
+///     - Must be for the same slot
+///     - Must have the same `shred_type`
+///     - Must both sigverify for the correct leader
+///     - If `shred1` and `shred2` share the same index they must be not equal
+///     - If `shred1` and `shred2` do not share the same index and are data shreds
+///       verify that they indicate an index conflict. One of them must be the
+///       LAST_SHRED_IN_SLOT, however the other shred must have a higher index.
+///     - If `shred1` and `shred2` do not share the same index and are coding shreds
+///       verify that they have conflicting erasure metas
 fn check_shreds<F>(leader_schedule: Option<F>, shred1: &Shred, shred2: &Shred) -> Result<(), Error>
```

Comment:
I took this approach because it seems more likely to detect the duplicate, as you only need 1 shred with index > than the previous last shred in slot. It is slightly less likely that you are able to receive both shreds that set the flag.
Makes the code more complicated, but I think it's the right thing to do here.

Reply:
> the same shred index with the same payload but different last_in_slot values

Shred flags which includes `last_in_slot` are encoded in the payload.
So if two shreds have different `last_in_slot` value they cannot have the same payload.